Add logic for edit/create custom procedures, wire up the playground

paulkaplan opened the pull request:
[work in progress, wanted to get it up so people could start taking a look]

This PR implements the `editProcedureCallback_` and `createProcedureCallback_` by wiring them to a common `externalProcedureDefCallback_`. This new function is just a stub, intended to be overwritten externally. An example implementation is then added to the custom procedure playground. 

![new-custom-procedure-playground](https://user-images.githubusercontent.com/654102/33146447-327c0650-cf92-11e7-972a-b73eea4ec750.gif)

I'd like to use this as a starting point for talking about how communication should work. Right now the external function gets the mutator from the procedure block of a given procedure, and a callback to call with an updated version of that mutator.

paulkaplan committed:
Add logic for edit/create custom procedures, wire up the playground

paulkaplan committed:
Move callbacks into callback factory.

rachel-fenichel commented on core/procedures.js:
```diff
@@ -358,9 +424,58 @@ Blockly.Procedures.editProcedureCallback_ = function(block) {
+
+/**
+ * Callback to create a new procedure custom command block.
+ * @private
```
Shouldn't be private, since you're explicitly telling people to override it.

rachel-fenichel commented on core/procedures.js:
```diff
@@ -306,40 +306,106 @@ Blockly.Procedures.mutateCallers = function(defBlock) {
 /**
  * Find the definition block for the named procedure.
- * @param {string} name Name of procedure.
+ * @param {string} procCode The identifier of the procedure to delete.
```
Nothing about this forces it to be a deletion.

rachel-fenichel commented on core/procedures.js:
```diff
@@ -306,40 +306,106 @@ Blockly.Procedures.mutateCallers = function(defBlock) {
  * @return {Blockly.Block} The procedure definition block, or null not found.
  */
-Blockly.Procedures.getDefinition = function(name, workspace) {
+Blockly.Procedures.getDefineBlock = function(procCode, workspace) {
```
This whole function can be replaced by finding the prototype block and then getting its parent.  Or vice versa.

rachel-fenichel commented on core/procedures.js:
```diff
@@ -358,9 +424,58 @@ Blockly.Procedures.editProcedureCallback_ = function(block) {
       return;
     }
     block = innerBlock;
+  } else if (!block.type == Blockly.PROCEDURES_PROTOTYPE_BLOCK_TYPE){
```
The above logic checks whether the right inputs, connections, etc. exists.  It can go into `getPrototypeBlock`.  Then this becomes

```
if (block.type == Blockly.PROCEDURES_DEFINITION_BLOCK_TYPE ||
    block.type == Blockly.PROCEDURES_CALL_BLOCK_TYPE) {
  block = Blockly.Procedures.getPrototypeBlock(block.getProcCode(), block.workspace)
}
```

paulkaplan replied:
Unfortunately the define block can't get the procCode, so we can't use getPrototypeBlock when all you have is the definition block.

rachel-fenichel replied:
Good point.

rachel-fenichel commented on core/procedures.js:
```diff
@@ -358,9 +424,58 @@ Blockly.Procedures.editProcedureCallback_ = function(block) {
+ */
+Blockly.Procedures.editProcedureCallbackFactory_ = function(block) {
+  return function(mutation) {
+    if (mutation) {
```
This belongs in Blockly.Procedures.mutateCallers: https://github.com/LLK/scratch-blocks/blob/develop/core/procedures.js#L311

paulkaplan committed:
Simplify block creation and definition and prototype lookups

rachel-fenichel commented on core/procedures.js:
```diff
@@ -358,9 +417,58 @@ Blockly.Procedures.editProcedureCallback_ = function(block) {
+      var defineBlock = Blockly.Procedures.getDefineBlock(block.getProcCode(),
+        block.workspace);
+      if (defineBlock) {
+        var callers = Blockly.Procedures.getCallers(block.getProcCode(),
```
This use of getCallers is kind of strained, since you had to go find the defineBlock even though you want to search in the entire workspace.  I added https://github.com/LLK/scratch-blocks/issues/1264 so we can circle back and clean it up later.

rachel-fenichel commented on core/procedures.js:
```diff
@@ -358,9 +417,58 @@ Blockly.Procedures.editProcedureCallback_ = function(block) {
+        alert('No define block on workspace'); // TODO decide what to do about this.
+      }
+      // And update the prototype block
+      block.domToMutation(mutation);
```
Does this also need an event?

rachel-fenichel commented on tests/custom_procedure_playground.html:
```diff
@@ -193,8 +204,14 @@
+      callback = null;
+      mutationRoot = null;
+      declarationWorkspace.clear();
+      definitionWorkspace.refreshToolboxSelection_()
```
No need to refresh the toolbox selection here.

paulkaplan replied:
That is done because the procedures category needs to be updated because of the change in blocks, and mutateCallers doesn't reach into the toolbox workspace to change those. Should it?

rachel-fenichel replied:
Good point.  This is fine.

rachel-fenichel commented on core/procedures.js:
```diff
@@ -306,40 +306,99 @@ Blockly.Procedures.mutateCallers = function(defBlock) {
+ * @return {Element} The mutation for a new custom procedure
+ */
+Blockly.Procedures.newProcedureMutation = function() {
+  var mutation = goog.dom.createDom('mutation');
```
This can also be done with textToDom, especially since it's always the same.

paulkaplan committed:
Update mutate callers.

Included updating the prototype block there so that the undos could all
be grouped together.

paulkaplan committed:
Update "public" annotation for external procedure def callback.

paulkaplan committed:
Fix comments

paulkaplan committed:
Use textToDom for creating default mutation.

paulkaplan commented:
@rachel-fenichel I updated this PR with more commits splitting out the logic into mutateCallers and whatnot. I tried to integrate some of the undo logic, but I think a more complete inspection of undoing should be undertaken separately.

rachel-fenichel commented on core/procedures.js:
```diff
@@ -358,9 +423,39 @@ Blockly.Procedures.editProcedureCallback_ = function(block) {
+  return function(mutation) {
+    if (mutation) {
+      Blockly.Procedures.mutateCallersAndPrototype(block.getProcCode(),
+        block.workspace, mutation);
```
Nit: indent +2

rachel-fenichel commented on core/procedures.js:
```diff
@@ -82,7 +82,7 @@ Blockly.Procedures.allProcedureMutations = function(root) {
   var mutations = [];
   for (var i = 0; i < blocks.length; i++) {
-    if (blocks[i].type === 'procedures_prototype') {
+    if (blocks[i].type == 'procedures_prototype') {
```
Use a constant for `procedures_prototype`.

rachel-fenichel commented on core/procedures.js:
```diff
@@ -276,70 +276,135 @@ Blockly.Procedures.getCallers = function(name, ws, definitionRoot,
+ * Find the prototype block for the named procedure.
+ * @param {string} procCode The identifier of the procedure.
+ * @param {!Blockly.Workspace} workspace The workspace to search.
+ * @return {Blockly.Block} The procedure prototype block, or null not found.
```
Annotate with `@private, @package, or @public`.  For this and all other functions you created.  Bonus points if you add annotations to functions that you touched.

rachel-fenichel commented on core/procedures.js:
```diff
@@ -276,70 +276,135 @@ Blockly.Procedures.getCallers = function(name, ws, definitionRoot,
+ */
+Blockly.Procedures.createProcedureDefCallback_ = function(workspace) {
+  Blockly.Procedures.externalProcedureDefCallback(
+    Blockly.Procedures.newProcedureMutation(),
```
Nit: indent +2, here and on the next lines.

rachel-fenichel replied:
Matching pattern three here: https://google.github.io/styleguide/jsguide.html#formatting-function-arguments

rachel-fenichel commented on core/procedures.js:
```diff
@@ -358,9 +423,39 @@ Blockly.Procedures.editProcedureCallback_ = function(block) {
+  } else if (block.type == Blockly.PROCEDURES_CALL_BLOCK_TYPE) {
+    // This is a call block, find the prototype corresponding to the procCode
+    block = Blockly.Procedures.getPrototypeBlock(
+      block.getProcCode(), block.workspace);
```
Nit: indentation.

rachel-fenichel commented on core/procedures.js:
```diff
@@ -358,9 +423,39 @@ Blockly.Procedures.editProcedureCallback_ = function(block) {
-      block.procCode_ + '")');
+  // Block now refers to the procedure prototype block, it is safe to proceed.
+  Blockly.Procedures.externalProcedureDefCallback(
+    block.mutationToDom(),
```
Nit: indentation.

rachel-fenichel reviewed (commented):
Overall looks good--I'm down to the nits on this review.

rachel-fenichel commented:
I agree that we should look into undo separately.  Tracked in https://github.com/LLK/scratch-blocks/issues/1194

The best way to approach events will be to come up with a set of test cases (sequences of actions) and then check that they can be undone and redone in a meaningful way.

paulkaplan committed:
Use constant for block type

paulkaplan committed:
Add jsdoc annotations

paulkaplan committed:
Fix indentation

paulkaplan commented:
Updated those nits, thanks @rachel-fenichel

rachel-fenichel reviewed (approved):
Ship it
